File: apps/fastapi/app/schemas/fee_collection.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field, conlist

PaymentMethod = Literal["CASH", "UPI", "BANK_TRANSFER"]


class FeePaymentAllocationCreate(BaseModel):
    fee_installment_id: int = Field(..., ge=1)
    amount_allocated: float = Field(..., gt=0)


class FeePaymentCollectRequest(BaseModel):
    student_id: int = Field(..., ge=1)
    tenant_id: Optional[int] = Field(None, ge=1)
    payment_method: PaymentMethod
    reference_no: Optional[str] = Field(None, max_length=100)
    notes: Optional[str] = None
    allocations: conlist(FeePaymentAllocationCreate, min_length=1)


class InvoicePaymentCollectRequest(BaseModel):
    invoice_id: int = Field(..., ge=1)
    tenant_id: Optional[int] = Field(None, ge=1)
    payment_amount: float = Field(..., gt=0)
    payment_method: PaymentMethod
    reference_no: Optional[str] = Field(None, max_length=100)
    payment_date: Optional[datetime] = None
    notes: Optional[str] = None
    bank_account_holder_name: Optional[str] = Field(None, max_length=100)
    bank_account_no: Optional[str] = Field(None, max_length=20)
    ifsc_code: Optional[str] = Field(None, max_length=11)


class FeePaymentCollectResponse(BaseModel):
    payment_id: int
    student_id: int
    total_amount: float
    payment_date: datetime
    receipt_number: Optional[str] = None


class FeeReceiptPaymentLineItem(BaseModel):
    sr_no: int
    txn_number: Optional[str] = None
    payment_type: str
    bank_name: Optional[str] = None
    amount: float


class FeeReceiptFeeDetailItem(BaseModel):
    sr_no: int
    fee_category_name: Optional[str] = None
    payable_for: Optional[str] = None
    amount: float
    paid_amount: float


class FeeReceiptDetailResponse(BaseModel):
    payment_id: int
    receipt_number: Optional[str] = None
    payment_date: datetime
    payment_method: str
    transaction_number: Optional[str] = None
    total_amount: float
    amount_in_words: str
    notes: Optional[str] = None
    student_name: str
    parent_name: Optional[str] = None
    admission_no: Optional[str] = None
    class_name: Optional[str] = None
    division_name: Optional[str] = None
    academic_year: Optional[str] = None
    invoice_no: Optional[str] = None
    installment: Optional[str] = None
    paid_for: Optional[str] = None
    created_by_name: Optional[str] = None
    payment_lines: list[FeeReceiptPaymentLineItem]
    fee_details: list[FeeReceiptFeeDetailItem] = []

File: apps/web/src/types/feeCollection.ts
export type PaymentMethod = "CASH" | "UPI" | "BANK_TRANSFER";

export interface FeePaymentAllocationCreate {
  fee_installment_id: number;
  amount_allocated: number;
}

export interface FeePaymentCollectRequest {
  student_id: number;
  tenant_id?: number;
  payment_method: PaymentMethod;
  reference_no?: string | null;
  notes?: string | null;
  allocations: FeePaymentAllocationCreate[];
}

export interface FeePaymentCollectResponse {
  payment_id: number;
  student_id: number;
  total_amount: number;
  payment_date: string;
  receipt_number?: string;
  payment_status?: string;
  pending_approval?: boolean;
}

export interface InvoicePaymentCollectRequest {
  invoice_id: number;
  tenant_id?: number;
  payment_amount: number;
  payment_method: PaymentMethod;
  reference_no?: string | null;
  payment_date?: string | null;
  notes?: string | null;
  bank_account_holder_name?: string | null;
  bank_account_no?: string | null;
  ifsc_code?: string | null;
}

export interface FeeReceiptPaymentLineItem {
  sr_no: number;
  txn_number?: string | null;
  payment_type: string;
  bank_name?: string | null;
  amount: number;
}

export interface FeeReceiptFeeDetailItem {
  sr_no: number;
  fee_category_name?: string | null;
  payable_for?: string | null;
  amount: number;
  paid_amount: number;
}

export interface FeeReceiptDetailResponse {
  payment_id: number;
  receipt_number?: string | null;
  payment_date: string;
  payment_method: string;
  transaction_number?: string | null;
  total_amount: number;
  amount_in_words: string;
  notes?: string | null;
  student_name: string;
  parent_name?: string | null;
  admission_no?: string | null;
  class_name?: string | null;
  division_name?: string | null;
  academic_year?: string | null;
  invoice_no?: string | null;
  installment?: string | null;
  paid_for?: string | null;
  created_by_name?: string | null;
  payment_lines: FeeReceiptPaymentLineItem[];
  fee_details: FeeReceiptFeeDetailItem[];
}

export type FeeApprovalStatus = "Pending Approval" | "Approved" | "Rejected" | "ALL";

export interface FeePaymentApprovalListItem {
  id: number;
  request_date: string;
  student_id: number;
  student_name: string;
  class_id?: number | null;
  class_name?: string | null;
  division_id?: number | null;
  division_name?: string | null;
  amount: number;
  payment_method: string;
  transaction_id?: string | null;
  status: FeeApprovalStatus;
}

export interface FeePaymentApprovalListResponse {
  items: FeePaymentApprovalListItem[];
  total: number;
  page: number;
  size: number;
}

export interface FeePaymentApprovalListParams {
  page?: number;
  size?: number;
  class_id?: number;
  division_id?: number;
  student_id?: number;
  status?: FeeApprovalStatus;
  search?: string;
}
